evm.md

```diff
@@ -2260,11 +2265,38 @@ There are several helpers for calculating gas (most of them also specified in th
     rule G*(GAVAIL, GLIMIT, REFUND) => GAVAIL +Int minInt((GLIMIT -Int GAVAIL)/Int 2, REFUND)
 
     syntax Int ::= #multComplexity(Int) [function]
- // ----------------------------------------------
+                 | #newMultComplexity(Int) [function]
+ // -------------------------------------------------
     rule #multComplexity(X) => X *Int X                                     requires X <=Int 64
     rule #multComplexity(X) => X *Int X /Int 4 +Int 96 *Int X -Int 3072     requires X >Int 64 andBool X <=Int 1024
     rule #multComplexity(X) => X *Int X /Int 16 +Int 480 *Int X -Int 199680 requires X >Int 1024
 
+    rule #newMultComplexity(X) => (X up/Int 8) ^Int 2
+
+    syntax Int ::= #calculateMultiplicationComplexity ( Int , Int )             [function]
+                 | #calculateIterationCount           ( Int , Int )             [function]
+                 | #calculateGasCost                  ( Int , Int , Int , Int ) [function]
+ // --------------------------------------------------------------------------------------
+    rule #calculateMultiplicationComplexity(BASELEN, MODLEN) => (maxInt(BASELEN, MODLEN) up/Int 8) ^Int 2
+
+    rule #calculateIterationCount(EXPLEN, EXPNT) => 1
+      requires EXPLEN <=Int 32 andBool EXPNT ==Int 0
+    rule #calculateIterationCount(EXPLEN, EXPNT) => maxInt(1, #bitLength(EXPNT) -Int 1)
+      requires EXPLEN <=Int 32 andBool notBool EXPNT ==Int 0
+    rule #calculateIterationCount(EXPLEN, EXPNT) => maxInt(1, ((8 *Int (EXPLEN -Int 32)) +Int (#bitLength(EXPNT &Int ((2 ^Int 256) -Int 1))) -Int 1 ))
+      requires EXPLEN >Int 32
+
+    rule #calculateGasCost(BASELEN, MODLEN, EXPLEN, EXPNT) => maxInt(200, #calculateMultiplicationComplexity(BASELEN, MODLEN) *Int #calculateIterationCount(EXPLEN, EXPNT) /Int 3)
+
+    syntax String ::= #toBinaryString ( Int , String ) [function]
+ // -------------------------------------------------------------
+    rule #toBinaryString(N, RESULT) => RESULT requires N ==Int 0
+    rule #toBinaryString(N, RESULT) => #toBinaryString(N /Int 2, Int2String(N %Int 2) +String RESULT) [owise]
+
+    syntax Int ::= #bitLength ( Int ) [function]
+ // --------------------------------------------
+    rule #bitLength(N) => lengthString(#toBinaryString(N, "":String))
```

Comment:
Pretty sure you can use `log2Int` instead of `#bitLength` here, round up the `log2Int` I think. It might be possible, maybe not.

But even if that doesn't work, you don't need ot use the `String` type, you can just keep track of a counter (cheaper memory wise).